### PR TITLE
feat(cmdline): extend it to support new evasions

### PR DESCRIPTION
**Pull request by fzipi**
Signed-off-by: Felipe Zipitria <felipe.zipitria@owasp.org>

This PR proposes a fix for #2632, where we recognize new evasion techniques.

In words, what we do here is:
- update the cmdline.py processor to include an extended pattern
- because now when the `cmdline unix` processor is applied, big lists will be generated, you need to be careful on the usage because it will generate massive lines
- rules 932100 and 932105 are removed, and two new rules are created instead: 932230 is for commands in 93210(0|5) with less than four characters (without the `@|~`), and 932240 is a generic rule that attempts to match the evasion techniques for words 
- tests from 932100 and 932105 were split into tests for 932230 and 932240 depending on the size of the input
- some white space cleanup in some files is also performed

### CODE REVIEW

File: util/regexp-assemble/data/932240.data [hunk 1]
```diff
@@ -0,0 +1,27 @@
+##! Please refer to the documentation at
+##! https://coreruleset.org/docs/development/regexp_assemble/.
+
+##! Rule 932240 (Generic RCE Unix command evasion detection)
+##!
+##! Conceptually, we want to match:
+##! - at least one char
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:7]:
```suggestion
##! - at least one character
```

File: util/regexp-assemble/data/932240.data [hunk 2]
```diff
@@ -0,0 +1,27 @@
...
+##! - at least one character from an evasion technique
+##! -
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:9]:
```suggestion

```

File: util/regexp-assemble/data/932240.data [hunk 3]
```diff
@@ -0,0 +1,27 @@
...
+##!+ i
+
+##!$ \b
+
+##! starting tokens prefix
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:14]:
```suggestion
```

File: util/regexp-assemble/data/932240.data [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+##!> assemble
+[a-z0-9_-]+
+##!=>
+
+##! This is the evasion detection coming from cmdline: [\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:19]:
```suggestion
##! This is the evasion detection from the cmdline processor: [\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?
```

File: util/regexp-assemble/data/932240.data [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+[\x5c'\"\[\]]+
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:20]: Add comment: what did we change here and why (w.r.t. to the cmdline processor).

File: util/regexp-assemble/data/932240.data [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+
+##! Rule 932240 (Generic RCE Unix command evasion detection)
+##!
+##! Conceptually, we want to match:
+##! - at least one char
+##! - at least one character from an evasion technique
+##! -
+##!+ i
+
+##!$ \b
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:12]: I don't think we need this.

```suggestion
##!$ \b
```

File: util/regexp-assemble/data/932240.data [hunk 7]
```diff
@@ -0,0 +1,27 @@
...
+##!> assemble
+[a-z0-9_-]+
+##!=>
+
+##! This is the evasion detection coming from cmdline: [\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?
+[\x5c'\"\[\]]+
+\$[\x5ca-z0-9_@?!#{*-]+
+##!=>
+
+[a-z0-9_-]+
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:24]: We need to address things like `10'000'000`. Maybe match more than x repetitions (x > 1)?

File: tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml [hunk 8]
```diff
@@ -0,0 +1,79 @@
...
+              Host: localhost
+              User-Agent: OWASP ModSecurity Core Rule Set
+            method: POST
+            port: 80
+            uri: /post
+            data: |
+              cmd=cu$@rl google.com
+            version: HTTP/1.1
+          output:
+            log_contains: id "932240"
```
Comment (theseion) [inline tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml:79]: Add test for `10'000'000`.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 9]
```diff
@@ -91,49 +91,23 @@ SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 1" "id:932012,phase:2,pass,nolog,skipAf
 # This is the base Rule to prevent Unix Command Injection
 # Please refer other rules 932105,932106,932101 to know more.
 #
-#  .932100 (base rule, PL1)
+#  .932230 (base rule, PL1, targets two and three chars commands)
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:94]:
```suggestion
#  .932230 (base rule, PL1, targets two and three character commands)
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 10]
```diff
@@ -91,49 +91,23 @@ SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 1" "id:932012,phase:2,pass,nolog,skipAf
...
...(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:[du][\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:\s|<|>).*|i[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?g|n[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?f)|i[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:d[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:\s|<|>).*|r[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?b|p)|x[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:z[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:\s|<|>).*|x[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?d)|w[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:3[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?m|h[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?o|c)|z[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:i[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?p|s[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?h)|7[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?z(?:[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?[ar])?|j[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?(?:j[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?s|q)|k[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?s[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?h|y[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?u[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?m)\b" \
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:104]: The new regex is really long, even with 2 and 3 characters only. Do we need to split into two rules?

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 11]
```diff
@@ -835,6 +808,42 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
     setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
     setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
 
+# -=[ Rule 932240 ]=-
+#
+# Generic RCE Bypass blocking using different techniques: see https://github.com/coreruleset/coreruleset/issues/2632
+#
+# This rule complements rule 932230 for commands of 4 characters and up, where we turn it into generic detection. Anything that uses a well-known evasion technique should be blocked at this level.
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:815]:
```suggestion
# This rule complements rule 932230 for commands of 4 characters and up, where we turn it into generic detection.
Anything that uses a well-known evasion technique should be blocked at this level.
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 12]
```diff
@@ -835,6 +808,42 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
...
+#
+# Regular expression generated from util/regexp-assemble/data/932220.data.
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:817]:
```suggestion
# Regular expression generated from util/regexp-assemble/data/932240.data.
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 13]
```diff
@@ -835,6 +808,42 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
...
+# To update the regular expression run the following shell script
+# (consult util/regexp-assemble/README.md for details):
+#   util/regexp-assemble/regexp-assemble.py update 932240
+#
+SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?i)[a-z0-9_-]+(?:\$[\x5ca-z0-9_@?!#{*-]+|[\x5c'\"\[\]]+)[a-z0-9_-]+\b" \
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:822]: Didn't we want to chain this rule???

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 14]
```diff
@@ -1033,7 +1042,7 @@ SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 3" "id:932016,phase:2,pass,nolog,skipAf
 #   util/regexp-assemble/regexp-assemble.py update 932106
 #
 # This rule is a stricter sibling of rule 932100.
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:1044]: Update mentions of 932100 in comments of other rules.

**Review by theseion — CHANGES_REQUESTED** [review]
Great stuff!

**Comment by RedXanadu** [comment]
Why do we want to remove rules? Sorry if there is an explanation, then I cannot find it.

**Comment by theseion** [comment]
Previously, the word list was split across two rules (932100 and 932105). We don't need the split anymore. We generally should not reuse rule IDs.

**Comment by dune73** [comment]
The new rule is very different from the original pair. That means existing Rule Exclusions might no longer be necessary. So better start with a new rule id.

File: util/regexp-assemble/lib/processors/cmdline.py [hunk 15]
```diff
@@ -104,7 +104,7 @@ def __init__(self, context: Context):
             context,
             # Insert these sequences between characters to prevent evasion.
             # This emulates the relevant parts of t:cmdLine.
-            r'''[\x5c'\"]*''',
+            r'''[\x5c'\"\[]*(?:\$[a-z0-9_@?!#{*-]*)?(?:\x5c)?''',
```
[the remark this answers is not in the document]
Comment (fzipi) [reply util/regexp-assemble/lib/processors/cmdline.py:110]: Added.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf (hunk 10 above)
Comment (fzipi) [reply rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:104]: It's 3K now. I wouldn't split it right now.....

File: tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml (hunk 8 above)
Comment (fzipi) [reply tests/regression/tests/REQUEST-932-APPLICATION-ATTACK-RCE/932240.yaml:79]: Added tests for this.

File: util/regexp-assemble/data/932240.data (hunk 7 above)
Comment (fzipi) [reply util/regexp-assemble/data/932240.data:22]: Added a test, and chained the rule so it doesn't match.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf (hunk 13 above)
Comment (fzipi) [reply rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]: Yes, it is chained now.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf (hunk 14 above)
Comment (fzipi) [reply rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf]: Removed.

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 16]
```diff
@@ -83,57 +83,31 @@ SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 1" "id:932012,phase:2,pass,nolog,skipAf
 # useful for this case. However, emulating the transformation makes
 # the regexp more complex.
 #
-# Regular expression generated from util/regexp-assemble/data/932100.data.
+# Regular expression generated from util/regexp-assemble/data/932230.data.
 # To update the regular expression run the following shell script
 # (consult util/regexp-assemble/README.md for details):
-#   util/regexp-assemble/regexp-assemble.py update 932100
+#   util/regexp-assemble/regexp-assemble.py update 9322300
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:89]:
```suggestion
#   util/regexp-assemble/regexp-assemble.py update 932230
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 17]
```diff
@@ -835,6 +805,45 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
     setvar:'tx.rce_score=+%{tx.critical_anomaly_score}',\
     setvar:'tx.inbound_anomaly_score_pl1=+%{tx.critical_anomaly_score}'"
 
+# -=[ Rule 932240 ]=-
+#
+# Generic RCE Bypass blocking using different techniques: see https://github.com/coreruleset/coreruleset/issues/2632
+#
+# This rule complements rule 932230 for commands of 4 characters and up, where we turn it into generic detection.
+# Anything that uses a well-known evasion technique should be blocked at this level.
+# The chained rule will exclude german numbering.
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:814]:
```suggestion
# The chained rule will exclude false positives due to german thousands separators (e.g., 10'000).
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 18]
```diff
@@ -835,6 +805,45 @@ SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAME
...
+    tag:'platform-unix',\
+    tag:'attack-rce',\
+    tag:'paranoia-level/1',\
+    tag:'OWASP_CRS',\
+    tag:'capec/1000/152/248/88',\
+    tag:'PCI/6.5.2',\
+    ver:'OWASP_CRS/4.0.0-rc1',\
+    severity:'CRITICAL',\
+    chain"
+    SecRule MATCHED_VAR "!@rx \'[0-9]" "t:none,\
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:840]:
```suggestion
    SecRule MATCHED_VAR "!@rx [0-9]\s*\'\s*[0-9]" \
        "t:none,\
```

File: rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf [hunk 19]
```diff
@@ -1025,15 +1034,16 @@ SecRule TX:DETECTION_PARANOIA_LEVEL "@lt 3" "id:932016,phase:2,pass,nolog,skipAf
...
+# For explanation of this rule, see rule 932230.
 #
 # Regular expression generated from util/regexp-assemble/data/932106.data.
 # To update the regular expression run the following shell script
 # (consult util/regexp-assemble/README.md for details):
 #   util/regexp-assemble/regexp-assemble.py update 932106
 #
-# This rule is a stricter sibling of rule 932100.
-SecRule REQUEST_COOKIES|!REQUEST_COOKIES:/__utm/|REQUEST_COOKIES_NAMES|ARGS_NAMES|ARGS|XML:/* "@rx (?:[;\n\r`]|t[\x5c'\"]*i[\x5c'\"]*m[\x5c'\"]*e|\$(?:\(?\(|{)|(?:\|)?\||\(\s*\)|[<>]\(|&?&|\{)\s*(?:(?:\w+=(?:[^\s]*|\$.*|\$.*|<.*|>.*|\'.*\'|\".*\")\s+|(?:\s*\(|!)\s*|\{|\$))*\s*(?:['\"])*(?:[\?\*\[\]\(\)\-\|+\w'\"\./\x5c]+/)?[\x5c'\"]*(?:(?:(?:a[\x5c'\"]*p[\x5c'\"]*t[\x5c'\"]*i[\x5c'\"]*t[\x5c'\"]*u[\x5c'\"]*d|u[\x5c'\"]*p[\x5c'\"]*2[\x5c'\"]*d[\x5c'\"]*a[\x5c'\"]*t)[\x5c'\"]*e|d[\x5c'\"]*n[\x5c'\"]*f|v[\x5c'\"]*i)[\x5c'\"]*(?:\s|<|>).*|p[\x5c'\"]*(?:a[\x5c'\"]*c[\x5c'\"]*m[\x5c'\"]*a[\x5c'\"]*n[\x5c'\"]*(?:\s|<|>).*|w[\x5c'\"]*d|s)|w[\x5c'\"]*(?:(?:\s|<|>).*|h[\x5c'\"]*o))\b" \
+# This rule was created as a stricter sibling of previous rule 932100.
```
Comment (theseion) [inline rules/REQUEST-932-APPLICATION-ATTACK-RCE.conf:1044]:
```suggestion
# This rule was created as a stricter sibling of rule 932100 (now replaced with 932230).
```

**Review by theseion — CHANGES_REQUESTED** [review]

File: util/regexp-assemble/data/932240.data (hunk 5 above)
Comment (fzipi) [reply util/regexp-assemble/data/932240.data:18]: Ok, did my best, considering the hour.

File: util/regexp-assemble/data/932240.data (hunk 6 above)
Comment (fzipi) [reply util/regexp-assemble/data/932240.data]: I've removed this, I wonder why is not showing.

File: util/regexp-assemble/data/932240.data [hunk 20]
```diff
@@ -10,11 +10,16 @@
 
 ##!+ i
 
+##! The previous cmdline evasion detection was based on the one from modsecurity: `[\x5c'\"\[]*`
```
Comment (theseion) [inline util/regexp-assemble/data/932240.data:13]:
```suggestion
##! The previous cmdline evasion detection was based on the one from modsecurity: `[\x5c'\"]*`
```

**Review by theseion — CHANGES_REQUESTED** [review]

**Review by theseion — APPROVED** [review]
❤️

**Pull request merged**